kore/src/Debug.hs

```diff
@@ -0,0 +1,581 @@
+
+instance (Debug a, Debug (f b)) => Debug (CofreeF f a b) where
+    debugPrec cofreeF =
+        -- Cannot have orphan instances of Generic and HasDatatypeInfo.
+        -- Use a fake instance instead.
+        debugPrecAux datatypeInfoCofreeF (debugSOP $ fromCofreeF cofreeF)
+
+datatypeInfoCofreeF :: DatatypeInfo '[ '[x, y] ]
+datatypeInfoCofreeF =
+    SOP.ADT
+        "Control.Comonad.Trans.Cofree"
+        "CofreeF"
+        (constrInfo :* Nil)
+  where
+    constrInfo = SOP.Infix ":<" SOP.RightAssociative 5
+
+fromCofreeF :: CofreeF f a b -> SOP I '[ '[a, f b] ]
+fromCofreeF (a :< fb) = SOP (Z (I a :* I fb :* Nil))
+
+instance
+    (Debug a, Debug (w (CofreeF f a (CofreeT f w a)))) =>
+    Debug (CofreeT f w a)
+  where
+    debugPrec x =
+        -- Cannot have orphan instances of Generic and HasDatatypeInfo.
+        -- Use a fake instance instead.
+        debugPrecAux datatypeInfoCofreeT (debugSOP (fromCofreeT x))
+
+datatypeInfoCofreeT :: DatatypeInfo '[ '[x] ]
+datatypeInfoCofreeT =
+    SOP.Newtype
+        "Control.Comonad.Trans.Cofree"
+        "CofreeT"
+        constrInfo
+  where
+    constrInfo = SOP.Record "CofreeT" (FieldInfo "runCofreeT" :* Nil)
+
+fromCofreeT :: CofreeT f w a -> SOP I '[ '[w (CofreeF f a (CofreeT f w a))] ]
+fromCofreeT (CofreeT x) = SOP (Z (I x :* Nil))
+
+instance (Debug k, Debug a) => Debug (Map.Map k a) where
+    debugPrec as precOut =
+        parens (precOut >= 10) ("Data.Map.fromList" <+> debug (Map.toList as))
+
+instance Debug a => Debug (Set a) where
+    debugPrec as precOut =
+        parens (precOut >= 10) ("Data.Set.fromList" <+> debug (Set.toList as))
+
+instance Debug a => Debug (Seq a) where
+    debugPrec as precOut =
+        parens (precOut >= 10)
+        $ "Data.Sequence.fromList" <+> debug (Foldable.toList as)
+
+instance Debug a => Debug (Const a b)
+
+instance Debug Bool
+
+instance (Debug a, Debug b) => Debug (Either a b)
+
+instance Debug ExitCode
+
+instance Debug GHC.CallStack
+
+instance Debug GHC.SrcLoc
+
+-- | Prints a typed hole for the function.
+instance (Typeable a, Typeable b) => Debug (a -> b) where
+    debugPrec f = \precOut ->
+        parens (precOut > 0)
+        $ "_" <+> "::" <+> (Pretty.pretty . show) (typeOf f)
+
+{- | 'Diff' displays the difference between values for debugging.
+
+@diff@ and @diffPrec@ should generally display the /minimal/ difference between
+two values, as far as it is practical to do so. Like 'debug', @diff@ and
+@diffPrec@ should produce valid Haskell source syntax to facilitate
+debugging. To elide the identical parts of two values, use holes (@_@).
+
+A default implementation is provided for @diffPrec@ by @diffPrecGeneric@, which
+only requires some instances to be derived:
+
+> data DataType = ...
+>     deriving (GHC.Generics.Generic)
+>
+> instance Generics.SOP.Generic DataType
+>
+> instance Generics.SOP.HasDatatypeInfo DataType
+>
+> instance Debug DataType
+>
+> instance Diff DataType
+
+ -}
+class Diff a where
```

Comment:
Not sure if it should be part of this PR, but I think that Id needs a custom `Diff` instance, similar to the custom `Eq` instance. I wonder if we have other custom `Eq` instances.

Reply:
I got around this by giving `AstLocation` a `Diff` instance which ignores all differences, because I think for debugging we would never care about that. But, I could give a custom instance for `Id` instead.